### ISSUE TITLE
[inductor] TorchInductor does not correctly recognize the grad status of model code

### ISSUE THREAD

**Issue by xuzhao9** [pytorch/pytorch#125474]
### 🐛 Describe the bug

At https://github.com/pytorch/pytorch/blob/30610251ec7b8f7e0507df06c3aadbcf90658e0e/torch/_inductor/compile_fx.py#L1371, torchinductor finds `torch.is_grad_enabled()` is `True` even if the model code sets `with torch.no_grad()`.

Reproduction script:

```
import torch
import torch.nn.functional as F

class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = torch.nn.Conv2d(1, 20, 5)

    def forward(self, x):
        return F.relu(self.conv1(x))

model = TestModule()
example_inputs = torch.randn([1, 1, 10, 10]).cpu()

@torch.compile
def run(model, example_inputs):
    with torch.no_grad():
        # Check inductor grad status at
        # https://github.com/pytorch/pytorch/blob/30610251ec7b8f7e0507df06c3aadbcf90658e0e/torch/_inductor/compile_fx.py#L1371
        # If this line is enabled, torchinductor believes torch.is_grad_enabled() is False
        # Otherwise, torchinductor believes torch.is_grad_enabled() is False
        # print("torch._inductor.config.cpp.weight_prepack", torch._inductor.config.cpp.weight_prepack)
        model(example_inputs)

run(model, example_inputs)
```

### Versions

The latest nightly release (20240503).

**Comment by xuzhao9** [comment]
This is reported by a torchbench user: https://github.com/pytorch/benchmark/issues/2253

**Comment by bdhirsh** [comment]
The simplest workaround is to move the `no_grad()` outside of your compiled region:

```
with torch.no_grad():
    run(model, example_inputs)
```

But I think we should be able to fix this. The problem is that your compiled region enables `no_grad` at the very beginning: so when `torch.compile` is invoked, we have not yet hit the `no_grad` region. When we actually trace the graph, though, we will find that none of the outputs of the graph require gradients.

AOTAutograd has some similar logic to recognize when it is compiling for inference vs training [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L572)

It's a bit annoying though, since inductor is choosing what fw_compiler to use **before** invoking AOTAutograd. So we would have to change the "fw_compiler" that we pass in from inductor, to dynamically switch between the inference or fw variant, depending on what AOTAutograd discovered.

**Comment by xuzhao9** [comment]
Thanks @bdhirsh for the response. I will keep this issue open since it is a valid issue with simple reproduction. On the torchbench side, we will workaround it by moving the `no_grad()` before the compile region.

**Comment by masnesral** [comment]
Issue scrubbing 2025Q1: Still an issue.

**Comment by masnesral** [comment]
This no longer repros for me on a recent build. Soooo fixed? @xuzhao9 do you want to verify?

**Comment by xuzhao9** [comment]
@masnesral  I can still reproduce with the original script on nightly 2.9.0.dev20250710+cu128, I updated the print statement location here:
https://github.com/pytorch/pytorch/blob/30610251ec7b8f7e0507df06c3aadbcf90658e0e/torch/_inductor/compile_fx.py#L1371

I am okay to close the issue since it is a rare use case and can be workarounded by wrapping `run()`.
